Address Windows build errors, most config-dependent.

ucko opened the pull request:
* In the absence of an external SQLGetPrivateProfileString, fail the build only when expecting to use an external driver manager, and expose tds_SQLGetPrivateProfileString to the odbc "connect" unit test. Don't let that test attempt to use SQLGetPrivateProfileStringW even in Unicode builds.
* src/odbc/win*.c: Accommodate static builds, with no hinstFreeTDS; in particular, never attempt to show FreeTDS dialog boxes there.
* src/utils/unittests/challenge.c: #include <freetds/sysdep_private.h>
  to ensure the availability of strcasecmp.

Split from #555.

freddy77 commented on src/odbc/connectparams.c:
```diff
@@ -556,7 +560,7 @@ odbc_build_connect_string(TDS_ERRS *errs, TDS_PARSED_PARAM *params, char **out)
 #if !HAVE_SQLGETPRIVATEPROFILESTRING
 
-#ifdef _WIN32
+#if defined(_WIN32)  &&  !defined(TDS_NO_DM)
```
Personally I would avoid this, Windows has a standard driver manager which should be used

ucko commented:
I acknowledge that bypassing Windows' native DM is unconventional, but it's logistically simpler to use driverless mode everywhere than to arrange to link this API's tests directly against it only in the absence of a system DM.  (The C\+\+ Toolkit's build system covers bundled projects itself rather than using their own build systems.)  I also seem to recall that the system DM insisted on sticking to drivers that had been registered system-wide, but perhaps that's not actually a problem, at least nowadays.

freddy77 commented:
Reasonable.

About the `connect.c` test... does it make sense to execute that part of the test that requires `get_entry` and so currently system `SQLGetPrivateProfileString`? It requires the driver to be registered as `FreeTDS` (see `DRIVER=FreeTDS;SERVER=%s;UID=%s;PWD=%s;DATABASE=%s;` connection string) which is only implemented by Driver Manager.

ucko commented:
AFAICT, that part of the test does actually succeed with this change in place, at least if I point the test at appropriate `freetds.conf` and `odbc.ini` files.  (I know there's logic to bypass that part when unable to look up the server name via `get_entry`, but it gets past that logic and proceeds with the remainder of that part.)

freddy77 commented on src/utils/unittests/challenge.c:
```diff
@@ -35,6 +35,7 @@
 #endif
 
 #include "tds_sysdep_public.h"
+#include <freetds/sysdep_private.h>
```
committed this hunk separately

ucko replied:
Thanks!

freddy77 commented on src/odbc/winlogin.c:
```diff
@@ -173,5 +173,9 @@ LoginDlgProc(HWND hDlg, UINT message, WPARAM wParam,	/* */
 bool
 get_login_info(HWND hwndParent, TDSLOGIN * login)
 {
+#ifdef DLL_EXPORT
```
Why this is `DLL_EXPORT` and not, for instance `TDS_NO_DM` ?

freddy77 replied:
Okay, I got it, if you are not a DLL you don't have the dialog to load. But then all code above is not used too.

freddy77 commented on src/odbc/winsetup.c:
```diff
@@ -428,6 +433,7 @@ check_installer_errors(void)
 	return SQL_SUCCEEDED(ret);
 }
 
+#ifdef DLL_EXPORT
```
yes, here it makes sense `DLL_EXPORT`, you cannot register something not an ODBC driver.

freddy77 commented on src/odbc/winsetup.c:
```diff
@@ -318,7 +318,12 @@ ConfigDSN(HWND hwndParent, WORD fRequest, LPCSTR lpszDriver, LPCSTR lpszAttribut
 	/* Maybe allow the user to edit it */
 	if (hwndParent && fRequest != ODBC_REMOVE_DSN) {
 		INT_PTR result;
+#ifdef DLL_EXPORT
```
I think all code in this file is exposing something for ODBC DM which cannot be used in case we are not a DLL.

ucko replied:
Further conditionalization may well be possible; I erred on the side of keeping more in.

freddy77 replied:
Merged the win* part.

freddy77 commented on src/odbc/unittests/connect.c:
```diff
@@ -12,18 +12,25 @@ init_connect(void)
 #ifdef _WIN32
 #include <odbcinst.h>
+#undef SQLGetPrivateProfileString
+#if !HAVE_SQLGETPRIVATEPROFILESTRING
+#  define SQLGetPrivateProfileString tds_SQLGetPrivateProfileString
+int tds_SQLGetPrivateProfileString(LPCSTR pszSection, LPCSTR pszEntry,
+                                   LPCSTR pszDefault, LPSTR pRetBuffer,
+                                   int nRetBuffer, LPCSTR pszFileName);
```
What happens if we use a DLL here? I would assume this test will refer to an undefined symbol and fail to load.

ucko replied:
Ah, yeah, (conditional) exporting is in order here.  However, I need a static build anyway for the sake of the `all_types`, `connection_string_parse`, and `utf8_4` tests, so our Windows builds leave this library static-only for simplicity.

freddy77 replied:
Tried to sort out this last hunk. Maybe code for that specific function could be just duplicated somehow? I tried to build for Windows and `SQLGetPrivateProfileString` from odbc system libraries was used, not the version from shared library.

ucko commented:
I've added CMake tuneups I'd missed earlier and rebased against `master`.  With these tuneups in place, Windows builds still naturally default to the system driver manager, but I can build the `connect` test against `tds_SQLGetPrivateProfileString` by editing `config.h` to comment out `HAVE_SQLGETPRIVATEPROFILESTRING`'s definition and instead predefine `TDS_NO_DM`.

freddy77 commented:
Merged
